### PR TITLE
Audio Control Application

### PR DESCRIPTION
**Pull request by baz2142**
## Description of changes

This PR adds the Ghaf Audio Control Appication to be able control AppVM's audio: speakers, microphones

This application uses Ghaf Audio Control library, which might be integrated in any bar, like Waybar or Ironbar, or into Ghaf Control Panel

## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [ ] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [ ] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [ ] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [x] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

## Testing

* Run application in taskbar "Audio Control"
* Check there are some sinks (audio devices)
* Check you can control sound by sliding a slider

### CODE REVIEW

**Comment by Mic92** [comment]
> Just heads up: #639 I can incorporate this into ironbar too.

You would need to rewrite this into Rust than.

File: overlays/custom-packages/waybar/default.nix [hunk 1]
```diff
@@ -4,11 +4,13 @@
...
-  prev.waybar.override {
-    hyprlandSupport = false;
-    swaySupport = false;
-    jackSupport = false;
-    cavaSupport = false;
-    pulseSupport = false;
-  }
-)
+  prev.waybar.overrideAttrs (_prevAttrs: {
+    patches = [./waybar-audio-control.patch];
```
[the remark this answers is not in the document]
Comment (baz2142) [reply overlays/custom-packages/waybar/default.nix]: Now it's an application inside Ghaf repo. Later I can integrate it into different places, like Waybar or Ironbar, or into Ghaf Control Panel

**Comment by baz2142** [comment]
> If we will include C++ code in this repo, should we also enable a C/C++ formatter with [treefmt](https://github.com/numtide/treefmt-nix). It seems to support clang-format.

Makes sense, actually. I will prepare a clang-format file

**Comment by milva-unikie** [comment]
Audio control app is not opening on Lenovo-X1. Clicking the icon also causes the screen to go black about 15 seconds later.

**Comment by leivos-unikie** [comment]
- Tested Audio Control app, launches, finds audio device, can control sound level and mute/unmute for integrated speakers/3.5mm headset. Control of microphone side is disabled for now (TBD?)
- All apps launch from the icons
- ci-test-automation passed and performance ok

**Comment by baz2142** [comment]
> If we will include C++ code in this repo, should we also enable a C/C++ formatter with [treefmt](https://github.com/numtide/treefmt-nix). It seems to support clang-format.

@humaidq-tii, I have enabled Clang Format support for treefmt and have put a .clang-format in my project

**Comment by baz2142** [comment]
> * Tested Audio Control app, launches, finds audio device, can control sound level and mute/unmute for integrated speakers/3.5mm headset. Control of microphone side is disabled for now (TBD?)
> * All apps launch from the icons
> * ci-test-automation passed and performance ok

@leivos-unikie, yeah, microphone will be enabled later in different PR

**Comment by josa41** [comment]
Although this does (mostly) work do we really need to build the control app ourselves?
I tested and most audio control apps do also work just fine.
I found one that is done in Rust (Not saying this is the best or even any good): https://github.com/Aurailus/Myxer

**Comment by mbssrc** [comment]
I would suggest to put this application into it's own repo in the TII space?

**Comment by baz2142** [comment]
> Although this does (mostly) work do we really need to build the control app ourselves? I tested and most audio control apps do also work just fine. I found one that is done in Rust (Not saying this is the best or even any good): https://github.com/Aurailus/Myxer

We can use a lot of apps, which supports PusleAudio and PipeWire, and they will work just fine for simple audio control. But here is "but": they are not adapted for Ghaf and not flexible:
* In our own app we can easily change protocol from PulseAudio to gRPC, DBus, Pipewire, whatever
* We have full control on UI and this is important: we don't have just sinks and sources, but we have different AppVM with their own icons, sets of apps, permissions and so on -- we might need to render it differently
* We can integrate it into Control Panel, make as an overlay and so. We can do whatever we want, we have a full control

I made the design of the app the way, that we can easily change protocol, UI and so on

**Comment by baz2142** [comment]
Close, as the app was put into ghafpkgs and enabled in ghaf by this PR:
https://github.com/tiiuae/ghaf/pull/808

**Pull request closed without merging**